datacube_stats/utils/tide_utility.py

```diff
@@ -0,0 +1,204 @@
+    tmp_lt = sorted(tide_dict.items(), key=lambda x: x[0])
+    dt_list = tmp_lt[1::3]
+    my_data = sorted(dt_list, key=itemgetter(1))
+    tmp_lt = [[tmp_lt[i+1][0].strftime("%Y-%m-%d"), 'ph']
+              if tmp_lt[i][1] < tmp_lt[i+1][1] and tmp_lt[i+2][1] < tmp_lt[i+1][1] else
+              [tmp_lt[i+1][0].strftime("%Y-%m-%d"), 'pl'] if tmp_lt[i][1] > tmp_lt[i+1][1] and
+              tmp_lt[i+2][1] > tmp_lt[i+1][1] else [tmp_lt[i+1][0].strftime("%Y-%m-%d"), 'f']
+              if tmp_lt[i][1] < tmp_lt[i+2][1] else [tmp_lt[i+1][0].strftime("%Y-%m-%d"), 'e']
+              for i in range(0, len(tmp_lt), 3)]
+    perc = 25 if per == 50 else per
+    max_height = my_data[-1][1]
+    min_height = my_data[0][1]
+    dr = max_height - min_height
+    lmr = min_height + dr*perc*0.01   # low tide max range
+    hlr = max_height - dr*perc*0.01   # high tide range
+    list_low = list()
+    list_high = list()
+    # doing for middle percentage to extract list of dates or for any percentage as per input.
+    if perc == 50:
+        list_high = sorted([[x[0].strftime('%Y-%m-%d'), x[1]]
+                            for x in my_data if (x[1] >= lmr) & (x[1] <= hlr) &
+                            (x[0] >= date_ranges[0][0]) & (x[0] <= date_ranges[0][1])])
+    else:
+        list_low = sorted([[x[0].strftime('%Y-%m-%d'), x[1]] for x in my_data if (x[1] <= lmr) &
+                          (x[0] >= date_ranges[0][0]) & (x[0] <= date_ranges[0][1])])
+        list_high = sorted([[x[0].strftime('%Y-%m-%d'), x[1]] for x in my_data if (x[1] >= hlr) &
+                           (x[0] >= date_ranges[0][0]) & (x[0] <= date_ranges[0][1])])
+    # Extract list of dates and type of tide phase within the date ranges for composite products
+    ebb_flow = str([tt for tt in tmp_lt
+                    if (datetime.strptime(tt[0], "%Y-%m-%d") >= date_ranges[0][0]) &
+                    (datetime.strptime(tt[0], "%Y-%m-%d") <= date_ranges[0][1])])
+    return dt_list, list_low, list_high, ebb_flow
+
+
+# This function is used to return a list of hydrological date range for dry wet geomedian
+# as per month list passed from config or by default from July to Nov.
+def get_hydrological_months(filter_product):
+    all_dates = list()
+    for k, v in filter_product['year'].items():
+        year = int(v)
+        months = filter_product['args'].get('months')
+        if months is not None:
+            st_dt = str(year+1)+str(months[0])+'01'
+            en_dt = str(year+1)+str(months[1])+'30'
+        else:
+            st_dt = '01/07/' + str(year+1)
+            en_dt = '30/11/' + str(year+1)
+        date_list = pd.date_range(st_dt, en_dt)
+        date_list = date_list.to_datetime().astype(str).tolist()
+        all_dates = all_dates + date_list
+    return all_dates
+
+
+# return a list of low high dates as per tide phase request
+def filter_sub_class(filter_product, list_low, list_high, ebb_flow):
```

Comment:
Could this class use some sort of data structure instead of all being specified in code? It looks really repetitive.

I'm also a bit dubious about `eval()` being used. What is stored in `ebb_flow` that requires it?